From b87ff74f7122eae55b76b1584e085d7eae972709 Mon Sep 17 00:00:00 2001
From: Jiani Liu <quic_jianil@quicinc.com>
Date: Thu, 6 Apr 2023 17:12:43 +0800
Subject: [PATCH] ath9k: fix ath9k p2pdev crash issue when chanctx is enabled

When chanctx(MCC) is enabled, ath9k will expose the APIs to support
P2P_DEVICE. The latter doesn't have any txq, it will caused kernel
crash when a p2pdev is added/removed.

Kernel log:
[31.355919] CPU 0 Unable to handle kernel paging request at virtual address
00000020, epc == 82f2d358, ra == 82f26ec4
...
[31.635640] Call Trace:
[31.638182] [<82f2d358>] ath_tx_node_init+0x4c/0xe0 [ath9k]
[31.643964] [<82f26ec4>] ath9k_set_txpower+0x510/0x53c [ath9k]
[31.650121] [<82d83208>] drv_add_interface+0x70/0x98 [mac80211]
[31.656308] [<82d9632c>] ieee80211_do_open+0x38c/0x918 [mac80211]
[31.662700] [<82d0c190>] cfg80211_check_station_change+0x7b8/0x1348 [cfg80211]

Change-Id: I16c1fcc18bf6dfe8c463950ef9c6cc08e2f5a704
Signed-off-by: Zhi Chen <zhichen@codeaurora.org>
Signed-off-by: Miaoqing Pan <miaoqing@codeaurora.org>
Signed-off-by: Jiani Liu <quic_jianil@quicinc.com>

diff --git a/drivers/net/wireless/ath/ath9k/main.c b/drivers/net/wireless/ath/ath9k/main.c
index 1a192ce..3578433 100644
--- a/drivers/net/wireless/ath/ath9k/main.c
+++ b/drivers/net/wireless/ath/ath9k/main.c
@@ -1362,7 +1362,9 @@ static int ath9k_add_interface(struct ieee80211_hw *hw,
 	an->sta = NULL;
 	an->vif = vif;
 	an->no_ps_filter = true;
-	ath_tx_node_init(sc, an);
+
+	if (vif->txq)
+		ath_tx_node_init(sc, an);
 
 	mutex_unlock(&sc->mutex);
 	return 0;
@@ -1425,7 +1427,8 @@ static void ath9k_remove_interface(struct ieee80211_hw *hw,
 	if (ath9k_uses_beacons(vif->type))
 		ath9k_beacon_remove_slot(sc, vif);
 
-	ath_tx_node_cleanup(sc, &avp->mcast_node);
+	if (vif->txq)
+		ath_tx_node_cleanup(sc, &avp->mcast_node);
 
 	ath9k_calculate_summary_state(sc, avp->chanctx);
 
-- 
2.7.4

